lib/Sema/TypeCheckDecl.cpp

```diff
@@ -2070,6 +2070,18 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
     }
   }
 
+  /// diagnoseSelfTypedParameters()
+  /// Diagnose instances of (Dynamic)Self as the type of a function arugument.
+  void diagnoseSelfTypedParameters(Type Ty, SourceLoc Loc) {
+    if (auto func = dyn_cast<FunctionType>(Ty->getCanonicalType()))
+      for (auto &param : func->getParams()) {
```

Comment:
I'm not sure the logic here is quite right. You're checking for a function type but calling this method with the type of a parameter.

Reply:
I wrote it that way so it could recurse. Perhaps the name needs improvement.